From 631b8035922a95739c34b650b88a139a8f16474a Mon Sep 17 00:00:00 2001
From: jason <jason@138bc75d-0d04-0410-961f-82ee72b054a4>
Date: Fri, 27 May 2011 20:48:24 +0000
Subject: [PATCH] 	PR c++/47277
 	* parser.c (cp_parser_unqualified_id): Don't check
 	constructor_name_p for enums.

git-svn-id: svn+ssh://gcc.gnu.org/svn/gcc/branches/gcc-4_6-branch@174363 138bc75d-0d04-0410-961f-82ee72b054a4

index 4d6ddcc..bc2b5bd 100644
--- a/gcc/cp/parser.c
+++ b/gcc/cp/parser.c
@@ -4320,7 +4320,8 @@ cp_parser_unqualified_id (cp_parser* parser,
 	    && (cp_lexer_peek_nth_token (parser->lexer, 2)->type
 		!= CPP_LESS)
 	    && (token->u.value == TYPE_IDENTIFIER (scope)
-		|| constructor_name_p (token->u.value, scope)))
+		|| (CLASS_TYPE_P (scope)
+		    && constructor_name_p (token->u.value, scope))))
 	  {
 	    cp_lexer_consume_token (parser->lexer);
 	    return build_nt (BIT_NOT_EXPR, scope);
new file mode 100644
index 0000000..5575ca6
--- /dev/null
+++ b/gcc/testsuite/g++.dg/cpp0x/enum18.C
@@ -0,0 +1,8 @@
+// PR c++/47277
+// { dg-options -std=c++0x }
+
+int main(void) {
+           enum e {};
+           e ev;
+           ev.e::~e_u();	// { dg-error "" }
+}
-- 
1.7.0.4

